akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventPusher.scala

```diff
@@ -66,18 +65,44 @@ private[akka] object EventPusher {
     implicit val ec: ExecutionContext = system.executionContext
     val protoAnySerialization = new ProtoAnySerialization(system)
 
-    def filterAndTransformFlow(filters: Future[immutable.Seq[proto.FilterCriteria]])
+    def filterAndTransformFlow(filters: Future[proto.ConsumerEventStart])
         : Flow[(EventEnvelope[Event], ProjectionContext), (ConsumeEventIn, ProjectionContext), NotUsed] =
       Flow
-        .futureFlow(filters.map { filterCriteria =>
-          val consumerFilter =
-            updateFilterFromProto(
-              Filter.empty(eps.settings.topicTagPrefix),
-              filterCriteria,
-              // FIXME not replicated, so maybe this is safe?
-              mapEntityIdToPidHandledByThisStream = identity)
+        .futureFlow(filters.map { startMessage =>
+          val (consumerFilter, replicatedEventOriginFilter) = {
+
+            startMessage.replicaInfo match {
+              case Some(replicaInfo) =>
+                val filter = updateFilterFromProto(
+                  Filter.empty(eps.settings.topicTagPrefix),
+                  startMessage.filter,
+                  mapEntityIdToPidHandledByThisStream = identity)
+                // needed to make sure we don't replicate events from the cloud back to cloud
+                val eventOriginFilterPredicate = eps.replicatedEventOriginFilter
+                  .getOrElse(throw new IllegalArgumentException(
+                    s"Entity ${eps.entityType} is a replicated entity but `replicatedEventOriginFilter` is not set"))
+                  .createFilter(replicaInfo)
+                val eventOriginFilterFlow =
+                  Flow[(EventEnvelope[Event], ProjectionContext)]
+                    .filter {
```

Comment:
This means that we don't store the offset. Wonder if that could cause some confusion? It's right that we don't have to emit them over grpc as FilteredEvent in this case, but maybe we should still pass it through the projection flow somehow? Maybe that is difficult and we could just handle it as any other FilteredEvent?

Reply:
Let me follow up with that after merging.

Reply:
That might be true. FilteredEvent is mostly? for transformation None, and I guess we don't have that for RES. Anyway, good to cover that case on the consumer side since we have the possibility in the protocol.